util/rowDecoder/decoder.go

```diff
@@ -134,3 +136,74 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, handle int
 	}
 	return row, nil
 }
+
+// SubstituteGenColsInDecodeColMap substitutes generated columns in every expression in decodeColMap
+// with non-generated one by looking up decodeColMap.
+func SubstituteGenColsInDecodeColMap(decodeColMap map[int64]Column) {
+	// Sort columns by ID in ascending order.
```

Comment:
Please Add comment why you need sort here. I think you have an assumption here.

Reply:
@crazycs520 I think line 150 does explain it. Do you have any suggestions?

Reply:
The reason is here: (https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html)

```
A generated column definition can refer to other generated columns, but only those occurring earlier in the table definition. A generated column definition can refer to any base (nongenerated) column in the table whether its definition occurs earlier or later.
```
Is it possible that the order of column ID is not the order of columns shown in the table definition? Using column ID is a bit risky here. @crazycs520 @tangenta

Reply:
@bb7133 You are right. I have found a test case that violates this assumption. I think 
 sorting by `table.Column.Offset` is the correct choice.

Reply:
> @bb7133 You are right. I have found a test case that violates this assumption. I think
> sorting by `table.Column.Offset` is the correct choice.

Good job, would you please show the case here?

Reply:
```sql
create table t (a int, c int as (a + 1));
insert into t (a) value (1);
alter table t add column b int as (a + 1) after a;
alter table t modify column c int as (b + 1);
alter table t add index idx(c);
```
Here the ID of column `a`, `b` and `c` is `1`, `3` and `2` respectively.

Reply:
It looks like you need fix bug:https://github.com/pingcap/tidb/issues/11365 first.